tests/functional/test_client.py

```diff
@@ -0,0 +1,198 @@
+        '--import',
+        os.path.abspath(key_file),
+    ]
+    subprocess.call(cmd)
+
+
+def create_dev_data(sdc_home, session_maker):
+    """
+    Based upon the functionality in the script, create_dev_data.py. This is
+    used to setup and configure the database and GPG keyring related metadata.
+    """
+    session = session_maker()
+    Base.metadata.create_all(bind=session.get_bind())
+    with open(os.path.join(sdc_home, Config.CONFIG_NAME), 'w') as f:
+        f.write(json.dumps({
+            'journalist_key_fingerprint': '65A1B5FF195B56353CC63DFFCC40EF1228271441',
+        }))
+    for reply_send_status in ReplySendStatusCodes:
+        try:
+            reply_status = session.query(ReplySendStatus).filter_by(
+                    name=reply_send_status.value).one()
+        except NoResultFound:
+            reply_status = ReplySendStatus(reply_send_status.value)
+            session.add(reply_status)
+            session.commit()
+
+
+def test_login_ensure_errors_displayed(qtbot, mocker):
+    """
+    We see an error if incomplete credentials are supplied to the login dialog.
+    """
+    w = Window()
+    login_dialog = LoginDialog(w)
+    login_dialog.show()
+    assert login_dialog.error_bar.error_status_bar.text() == ""
+    qtbot.keyClicks(login_dialog.username_field, "journalist")
+    qtbot.mouseClick(login_dialog.submit, Qt.LeftButton)
+    expected = "Please enter a username, password and two-factor code."
+    actual = login_dialog.error_bar.error_status_bar.text()
+    assert actual == expected
+
+
+@pytest.mark.vcr()  # Ensure any API network traffic is recorded/replayed.
+def test_login_as_journalist(qtbot, mocker):
+    """
+    The app is visible if the user logs in with apparently correct credentials.
+    """
+    # Once out of scope, is deleted.
+    tempdir = get_safe_tempdir()
+    # Create a clean context.
+    gui, controller = get_test_context(tempdir)
+    gui.setup(controller)
+    # Fill in UI with good credentials.
+    qtbot.keyClicks(gui.login_dialog.username_field, USERNAME)
+    qtbot.keyClicks(gui.login_dialog.password_field, PASSWORD)
+    qtbot.keyClicks(gui.login_dialog.tfa_field, "493941")
+    # The waitSignal context handler is used to allow the API thread to call
+    # and then (ultimately) emit the expected signal. This pattern will need to
+    # be used with all API calls. For more information about this method, see:
+    # https://pytest-qt.readthedocs.io/en/latest/signals.html
+    with qtbot.waitSignal(controller.authentication_state, timeout=10000):
+        qtbot.mouseClick(gui.login_dialog.submit, Qt.LeftButton)
+    # The main window is visible (indicating a successful login).
+    assert gui.isVisible()
+    # The login box no longer exists.
+    assert gui.login_dialog is None
+
+
+@pytest.mark.vcr()
+def test_logout_as_journalist(qtbot, mocker):
+    """
+    A journalist can successfully log out of the application.
+    """
+    totp = "670099"
```

Comment:
We need explanation (may be at the top + developer documentation) on how to get the `totp` value while writing or updating a test case. And how to keep using it after the test data is finalized.

Reply:
This is already in the docstring at the top of the module.